src/hal/src/pso/mod.rs

```diff
@@ -213,40 +239,77 @@ pub enum BasePipeline<'a, P: 'a> {
 /// More importantly, they are fast to execute, since the driver
 /// can optimize out the branch on that other PSO creation.
 #[derive(Debug, Clone)]
-pub struct SpecializationConstant {
+pub struct RawSpecializationConstant {
     /// Constant identifier in shader source.
     pub id: u32,
     /// Value to override specialization constant.
     pub range: Range<u16>,
 }
 
 /// Specialization information structure.
-#[derive(Debug, Copy)]
-pub struct Specialization<'a> {
+#[derive(Debug, Copy, Clone)]
+pub struct RawSpecialization<'a> {
     /// Constant array.
-    pub constants: &'a [SpecializationConstant],
+    pub constants: &'a [RawSpecializationConstant],
     /// Raw data.
     pub data: &'a [u8],
 }
 
-impl<'a> Default for Specialization<'a> {
+impl<'a> Default for RawSpecialization<'a> {
     fn default() -> Self {
-        Specialization {
+        RawSpecialization {
             constants: &[],
             data: &[],
         }
     }
 }
 
-impl<'a> Clone for Specialization<'a> {
-    fn clone(&self) -> Self {
-        Specialization {
-            constants: self.constants,
-            data: self.data,
+/// Temporary storage for specialization constants.
+pub struct SpecializationStorage {
+    data: Vec<u8>,
+    constants: Vec<RawSpecializationConstant>,
+}
+
+/// Convert typed specialization into untyped one.
+pub trait Specialization<'a> {
+    /// Build `Specialization` value.
+    fn into_raw(&self, storage: &'a mut SpecializationStorage) -> RawSpecialization<'a>;
```

Comment:
what role does `SpecializationStorage` play here?

Reply:
It owns what `RawSpecialization`, borrows.